frontend/src/metabase/visualizations/echarts/cartesian/option/series.ts

```diff
@@ -0,0 +1,159 @@
+import type { EChartsOption, RegisteredSeriesOption } from "echarts";
+import type { SeriesLabelOption } from "echarts/types/src/util/types";
+import type {
+  SeriesModel,
+  CartesianChartModel,
+} from "metabase/visualizations/echarts/cartesian/model/types";
+import type {
+  ComputedVisualizationSettings,
+  RenderingContext,
+} from "metabase/visualizations/types";
+import type { SeriesSettings } from "metabase-types/api";
+import { isNotNull } from "metabase/lib/types";
+
+const buildEChartsLabelOptions = (
+  seriesModel: SeriesModel,
+  settings: ComputedVisualizationSettings,
+  { getColor, fontFamily, formatValue }: RenderingContext,
+): SeriesLabelOption => {
+  const valueFormatter = (value: unknown) =>
+    formatValue(value, {
+      ...(settings.column?.(seriesModel.column) ?? {}),
+      jsx: false,
+      compact: settings["graph.label_value_formatting"] === "compact",
+    });
+
+  return {
+    show: settings["graph.show_values"],
+    position: "top",
+    fontFamily,
+    fontWeight: 900,
+    fontSize: 12,
+    color: getColor("text-dark"),
+    textBorderColor: getColor("white"),
+    textBorderWidth: 3,
+    formatter: datum => {
+      const dimensionIndex = datum?.encode?.y[0];
+      const dimensionName =
+        dimensionIndex != null ? datum?.dimensionNames?.[dimensionIndex] : null;
+      if (dimensionIndex == null || dimensionName == null) {
+        return " ";
+      }
+      const value = (datum?.value as any)?.[dimensionName];
+      return valueFormatter(value);
+    },
+  };
+};
+
+const buildEChartsBarSeries = (
+  seriesModel: SeriesModel,
+  seriesSettings: SeriesSettings,
+  seriesColor: string,
+  settings: ComputedVisualizationSettings,
+  dimensionDataKey: string,
+  yAxisIndex: number,
+  renderingContext: RenderingContext,
+): RegisteredSeriesOption["bar"] => {
+  const stackName =
+    settings["stackable.stack_type"] != null ? `bar_${yAxisIndex}` : undefined;
+
+  return {
+    type: "bar",
+    yAxisIndex,
+    stack: stackName,
+    encode: {
+      y: seriesModel.dataKey,
+      x: dimensionDataKey,
+    },
+    label: buildEChartsLabelOptions(seriesModel, settings, renderingContext),
+    labelLayout: {
+      hideOverlap: settings["graph.label_value_frequency"] === "fit",
+    },
+    itemStyle: {
+      color: seriesColor,
+    },
+  };
+};
+
+const buildEChartsLineAreaSeries = (
+  seriesModel: SeriesModel,
+  seriesSettings: SeriesSettings,
+  seriesColor: string,
+  settings: ComputedVisualizationSettings,
+  dimensionDataKey: string,
+  yAxisIndex: number,
+  renderingContext: RenderingContext,
+): RegisteredSeriesOption["line"] => {
+  const display = seriesSettings?.display ?? "line";
```

Comment:
At this point `seriesSetting.display` should be defined because we get the computed value from the `settings.series` function call on line 124. Maybe we should throw an error here instead, because if there is no default display value on series settings something must have went wrong.